src/tracing/agent.h

```diff
@@ -31,46 +32,67 @@ class TracingController : public v8::platform::tracing::TracingController {
   }
 };
 
+class Agent;
 
-class Agent {
+// Allows agent to be destroyed early. Methods are noop after the agent was
+// destroyed
+class AgentHandle {
  public:
-  // Resetting the pointer disconnects client
-  using ClientHandle = std::unique_ptr<std::pair<Agent*, int>,
-                                       void (*)(std::pair<Agent*, int>*)>;
+  explicit AgentHandle(Agent* agent);
+  void Reset();
+  void AppendTraceEvent(TraceObject* trace_event);
+  void Flush(bool blocking);
+  void Disconnect(int client);
 
-  static ClientHandle EmptyClientHandle() {
-    return ClientHandle(nullptr, DisconnectClient);
-  }
+ private:
+  Mutex mutex_;
+  Agent* agent_;
+};
+
+class ClientHandle {
+ public:
+  ClientHandle(std::shared_ptr<AgentHandle> agent, int id)
+               : agent_(agent), id_(id) {}
+  ClientHandle(const ClientHandle&) = delete;
+  ClientHandle& operator=(const ClientHandle&) = delete;
+  ~ClientHandle();
+  int id() const { return id_; }
+
+ private:
+  std::shared_ptr<AgentHandle> agent_;
+  const int id_;
+};
+
+class Agent {
+ public:
   explicit Agent(const std::string& log_file_pattern);
+  ~Agent();
   void Stop();
 
   TracingController* GetTracingController() { return tracing_controller_; }
 
   // Destroying the handle disconnects the client
-  ClientHandle AddClient(const std::set<std::string>& categories,
-                         std::unique_ptr<AsyncTraceWriter> writer);
+  std::unique_ptr<ClientHandle> AddClient(
+      const std::set<std::string>& categories,
+      std::unique_ptr<AsyncTraceWriter> writer);
 
   // These 3 methods operate on a "default" client, e.g. the file writer
   void Enable(const std::string& categories);
   void Enable(const std::set<std::string>& categories);
   void Disable(const std::set<std::string>& categories);
   std::string GetEnabledCategories();
 
-  void AppendTraceEvent(TraceObject* trace_event);
+  // void AppendTraceEvent(TraceObject* trace_event);
```

Comment:
Should this be removed rather than commented out?

Reply:
I pushed a wrong version - this method is still in use...